modules/biobank/php/specimendao.class.inc

```diff
@@ -0,0 +1,719 @@
+                $pA[$row['protocolId']][$row['attributeId']]['required']   = $row['required'];
+            }
+        }
+
+        return $pA;
+    }
+
+    /**
+     * Queries all rows from the biobank_specimen_method_attribute_rel table
+     * and returns a nested array of specimen-method-specific attribute data
+     * with the Method ID as the first index, and the Attribute ID as the
+     * second index.
+     *
+     * @return array $mA A nested array of attribute data
+     *                   indexed by Method ID and Attribute
+     *                   ID
+     */
+    public function getMethodAttributes() : array
+    {
+        $query  = "SELECT bsm.SpecimenMethodID as methodId,
+                         bsa.SpecimenAttributeID as attributeId,
+                         bsa.Label as name,
+                         bsa.DatatypeID as datatypeId,
+                         bsa.ReferenceTableID as refTableId,
+                         bsma.Required as required
+                  FROM biobank_specimen_method_attribute_rel bsma
+                  JOIN biobank_specimen_method bsm
+                    USING (SpecimenMethodID)
+                  JOIN biobank_specimen_attribute bsa
+                    USING (SpecimenAttributeID)";
+        $result = $this->db->pselect($query, array());
+        $mA     = array(); //methodAttributes
+        foreach ($result as $row) {
+            foreach ($row as $value) {
+                $mA[$row['methodId']][$row['attributeId']]['name']       = $row['name'];
+                $mA[$row['methodId']][$row['attributeId']]['datatypeId'] = $row['datatypeId'];
+                $mA[$row['methodId']][$row['attributeId']]['refTableId'] = $row['refTableId'];
+                $mA[$row['methodId']][$row['attributeId']]['required']   = $row['required'];
+            }
+        }
+
+        return $mA;
+    }
+
+    /**
+     * This function receives a Specimen Object, converts it into a SQL format
+     * and inserts it into all relevant tables in the database.
+     *
+     * @param object $specimen The Specimen Object to be saved the
+     *                         database.
+     *
+     * @return void
+     */
+    public function saveInstance(Specimen $specimen) : void
+    {
+        // Converts Specimen Object into SQL format.
+        $specimenArray = $this->getArrayFromInstance($specimen);
+        $data          = $this->getSQLFromArray($specimenArray);
+
+        // TODO: Look into implementing a transaction here
+        // $this->db->_PDO->beginTransaction();
+        // $this->db->_PDO->commit();
+        // $this->db->_PDO->rollBack();
+
+        // Inserts or Updates the biobank_specimen table with respective data.
+        $this->db->insertOnDuplicateUpdate(
+            'biobank_specimen',
+            $data['biobank_specimen']
+        );
+
+        $specimenId = $specimen->getId();
+        if (isset($specimenId)) {
+            // If update:
+            // XXX: Logic will likely eventually go here.
```

Comment:
Hmm, you're right. I guess nothing wil really go in here in this `if` clause. The `else` clause is important though, because it's an insert, then the new `specimenId` that was just inserted needs to be associated with all the other inserts. I guess this could be done regardless of whether it's an insert or update, it just becomes redundant if it's an update.

Reply:
Actually, after looking up `getLastInsertId`, I realised I do need to have this check. If I try and run `getLastInsertId` after an update, I wont get the proper `id` that I want because it wasn't an insert. If you look at the containerdao there's some logic that goes here.

Reply:
well 2 things here
 - the empty if is not necessary unless you are actually putting logic in there (its fine if ur using it in the container, leave it there)
 - the comment saying `logic will likely...` is misleading